### PR TITLE
Support tunnel routing in Multi-Pool IPAM mode

### PR DESCRIPTION
**Pull request by pippolo84**
Rework of https://github.com/cilium/cilium/pull/38015 after the introduction of clusterID aware ipcache in https://github.com/cilium/cilium/pull/38379

The main difference with https://github.com/cilium/cilium/pull/38015 is commit https://github.com/cilium/cilium/pull/38483/commits/fa35fa0247909ada83f72952daef85a284bc1f40 where we take advantage of the `ClusterPrefix` parameter type in ipcache API to enrich the node prefixes (namely, the CiliumInternalIP, the Ingress and health IPs and the v4/v6 allocation CIDRs) with the result of a customizable PrefixClusterMutatorFn. This guarantees feature parity with the  tunnel map previously used for node allocation CIDRs.

Description of the previous version of the PR (https://github.com/cilium/cilium/pull/38015) below:

---

This PR introduces support for tunnel routing with the Multi-Pool IPAM mode. This is achieved by removing the reliance on the tunnel map as a fallback mechanism and instead use Pod CIDR entries in IPCache that act as this fallback. The pod CIDR entries contain the node's tunnel endpoint and encryption key and have the `reserved:world` identity (i.e. the same identity as the existing catch-all fallback, which would have been hit before this PR). 

The bulk of the work in this PR was done by @pippolo84. This PR is a replacement for his PR https://github.com/cilium/cilium/pull/37146 with a few notable changes:

1. The first commit has been reworked to add unit tests and fix up some issues in the previous version (c.f. https://github.com/cilium/cilium/pull/37146#discussion_r1973593637).
2. A commit has been added adding some unit tests to the IPCache package emulating some of the expected corner cases.
3. Commit `node: Remove insertions and deletions to tunnel map` has been dropped. In other words, even though the tunnel map is no longer used in the `main` branch, we still populate it. This addresses concerns with regards to downgrades.
4. The Conformance Multi-Pool workflow has been extended to also test in tunnel mode.

This change will eventually allow us to remove the tunnel map completely (ref https://github.com/cilium/cilium/issues/20170).

---

```release-note
Add support for tunnel routing in multi-pool IPAM mode
```

### CODE REVIEW

File: pkg/node/manager/manager.go [hunk 1]
```diff
@@ -987,13 +1018,47 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
...
+			prefix, ok := netipx.FromStdIPNet(podCIDR.IPNet)
+			if !ok {
+				continue
+			}
+
+			metadata := []ipcache.IPMetadata{
+				worldLabelForPrefix(prefix),
+				ipcacheTypes.TunnelPeer{Addr: tunnelIP},
+			}
+			if m.nodeAddressHasEncryptKey() {
```
Comment (giorio94) [inline pkg/node/manager/manager.go]: The EncryptKey should be always configured here. `nodeAddressHasEncryptKey` applies to node addresses only, where it should be set only if node-to-node encryption is enabled.

File: pkg/node/manager/manager.go [hunk 2]
```diff
@@ -877,6 +889,25 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
...
+		mu := make([]ipcache.MU, 0, len(ipv4PodCIDRs)+len(ipv6PodCIDRs))
+		for entry := range m.podCIDREntries(n.Source, resource, ipv4PodCIDRs, nodeIP, n.EncryptionKey) {
+			mu = append(mu, entry)
+			podCIDRsAdded = append(podCIDRsAdded, entry.Prefix.AsPrefix())
+		}
+		for entry := range m.podCIDREntries(n.Source, resource, ipv6PodCIDRs, nodeIP, n.EncryptionKey) {
+			mu = append(mu, entry)
+			podCIDRsAdded = append(podCIDRsAdded, entry.Prefix.AsPrefix())
+		}
+		m.ipcache.UpsertMetadataBatch(mu...)
```
Comment (giorio94) [inline pkg/node/manager/manager.go:918]: This problem potentially also exists for the other async ipcache upsertions performed here (and this is likely not the best venue to fix it), and it lies in the fact that we are not waiting on the updates to be actually propagated down to the ipcache map. However, mentioning here as somewhat related, and that could theoretically lead to drops upon endpoint regeneration in case the ipcache is not fully populated yet (probably not caused by the missing fallback CIDRs though, as they should serve as a fallback only). 

AFAIU, we do wait on the first revision during daemon startup \[1]. Yet, while at that point we are guaranteed to have already synchronized with Kubernetes, no such guarantee exists for the kvstore and/or clustermesh (while we do block endpoint regeneration waiting on them a bit later), and a part of the initial node update events could be received afterwards as well. That said, simply waiting here would not work either (and go against the async principle), because the ipcache may not even have been started yet.

This is just to say that we should likely uniform the wait for sync process, and ensure that we wait for the first ipcache revision only after ensuring that all possible sources are synchronized. This is the more important the more subsystems we migrate to the asynchronous ipcache.

\[1]: https://github.com/cilium/cilium/blob/3a382254dab57eccd2bc5fb28de4537d7395d4be/daemon/cmd/daemon_main.go#L1694-L1696

File: daemon/cmd/daemon_main.go [hunk 3]
```diff
@@ -1330,10 +1330,6 @@ func initEnv(vp *viper.Viper) {
 		option.Config.EncryptInterface = append(option.Config.EncryptInterface, link)
 	}
 
-	if option.Config.TunnelingEnabled() && option.Config.EnableAutoDirectRouting {
-		log.Fatalf("%s cannot be used with tunneling. Packets must be routed through the tunnel device.", option.EnableAutoDirectRoutingName)
-	}
```
Comment (giorio94) [inline daemon/cmd/daemon_main.go]: This removal doesn't look related to me.

File: .github/workflows/conformance-multi-pool.yaml [hunk 4]
```diff
@@ -81,6 +81,15 @@ jobs:
           sha: ${{ inputs.SHA || github.sha }}
 
   multi-pool-ipam-conformance-test:
+    strategy:
```
Comment (giorio94) [inline .github/workflows/conformance-multi-pool.yaml:101]: Nit. Given that this is now a matrix, it could be worthy moving the "Wait for images" step outside. E.g., \[1]

\[1]: https://github.com/cilium/cilium/blob/ade6c0c6a0ff3a279d0002722bd6c0073bcc6d9c/.github/workflows/tests-e2e-upgrade.yaml#L81-L94

File: pkg/node/manager/manager.go [hunk 5]
```diff
@@ -818,10 +822,16 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
 	var nodeIPsAdded, healthIPsAdded, ingressIPsAdded, podCIDRsAdded []netip.Prefix
 
 	for _, address := range n.IPAddresses {
-		prefixCluster := cmtypes.NewLocalPrefixCluster(ip.IPToNetPrefix(address.IP))
+		prefix := ip.IPToNetPrefix(address.IP)
+		var prefixCluster cmtypes.PrefixCluster
+		if address.Type == addressing.NodeCiliumInternalIP {
+			prefixCluster = cmtypes.PrefixClusterFrom(prefix.Addr(), prefix.Bits(), m.prefixClusterMutatorFn(&n)...)
```
Comment (giorio94) [inline pkg/node/manager/manager.go]: Nit. `cmtypes.PrefixClusterFrom` could be likely changed to directly take a prefix as parameter, rather than address and bits.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Thanks! A couple of comments inline, besides that looks great to me.

**Review by ghost — APPROVED** [review]
docs good

**Review by Artyop — APPROVED** [review]
besides Marco's review about waiting for image step, looks good for github-sec and ci-structure

File: pkg/node/manager/manager.go (hunk 5 above)
Comment (pippolo84) [reply pkg/node/manager/manager.go]: Nice idea :+1:

File: .github/workflows/conformance-multi-pool.yaml (hunk 4 above)
Comment (giorio94) [reply .github/workflows/conformance-multi-pool.yaml:101]: The `wait-for-images` job should be added to the list of `needs` of `multi-pool-ipam-conformance-test`.

**Review by giorio94 — APPROVED** [review]
One more comment inline (above), looks good to me otherwise.

File: .github/workflows/conformance-multi-pool.yaml (hunk 4 above)
Comment (pippolo84) [reply .github/workflows/conformance-multi-pool.yaml:101]: Updated, thanks for the heads up!

**Review by doniacld — APPROVED** [review]

File: pkg/node/manager/manager.go (hunk 2 above)
Comment (pippolo84) [reply pkg/node/manager/manager.go:918]: Opened https://github.com/cilium/cilium/issues/38520 to discuss this further and clarify the possible impact.

**Review by joamaki — APPROVED** [review]

**Comment by pippolo84** [comment]
So, it seems that the failing Ginkgo test is not propagating CEPs to simulate an unknown endpoint. But now that we insert pod CIDRs into ipcache, alongside the encryption key, the traffic is encrypted (pod CIDR entry in the ipcache includes the pod IP) and it goes through, while the test was expecting a drop.
We can probably drop the old Gingko test in favor of https://github.com/cilium/cilium/pull/35231

**Comment by pippolo84** [comment]
Blocked on https://github.com/cilium/cilium/pull/38538
